pkg/resource/default.go

```diff
@@ -0,0 +1,121 @@
+
+	corev1alpha1 "github.com/crossplaneio/crossplane/pkg/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane/pkg/logging"
+)
+
+const (
+	controllerNameDefault = "defaultclass.crossplane.io"
+)
+
+var logDefault = logging.Logger.WithName("controller").WithValues("controller", controllerNameDefault)
+
+// Error strings
+const (
+	errNoDefault        = "unable to locate a default resource class for claim kind"
+	errMultipleDefaults = "multiple default classes defined for claim kind"
+	errInvalidLabel     = "unable to set label for claim kind"
+)
+
+// DefaultReconciler reconciles resource claims to the
+// default resource class for their given kind. Predicates
+// ensure that only claims with no resource class reference
+// are reconciled.
+type DefaultReconciler struct {
+	client   client.Client
+	newClaim func() Claim
+}
+
+// NewDefaultReconciler creates a new DefaultReconciler for the claim kind
+func NewDefaultReconciler(m manager.Manager, of ClaimKind) *DefaultReconciler {
+	nc := func() Claim { return MustCreateObject(schema.GroupVersionKind(of), m.GetScheme()).(Claim) }
+
+	_ = nc()
+
+	r := &DefaultReconciler{
+		client:   m.GetClient(),
+		newClaim: nc,
+	}
+
+	return r
+}
+
+// Reconcile reconciles a claim to the default class reference for its kind
+func (r *DefaultReconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	logDefault.V(logging.Debug).Info("Reconciling", "request", req)
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	claim := r.newClaim()
+	if err := r.client.Get(ctx, req.NamespacedName, claim); err != nil {
+		// There's no need to requeue if we no longer exist. Otherwise we'll be
+		// requeued implicitly because we return an error.
+		return reconcile.Result{}, errors.Wrap(IgnoreNotFound(err), errGetClaim)
+	}
+
+	// Get claim group and kind information
+	claimKind := claim.GetObjectKind()
+	claimGroupVersionKind := claimKind.GroupVersionKind()
+	claimGroupKind := claimGroupVersionKind.GroupKind()
+	claimGroupKindStr := strings.ToLower(claimGroupKind.String())
+
+	// Create label selector in crossplane-system namespace
+	options := client.InNamespace("crossplane-system")
+	if err := options.SetLabelSelector(claimGroupKindStr + "/default=" + "true"); err != nil {
+		return reconcile.Result{Requeue: false}, errors.Wrap(IgnoreNotFound(r.client.Status().Update(ctx, claim)), errInvalidLabel)
+	}
+
+	// Get resource classes with claim kind as default
+	classes := &corev1alpha1.ResourceClassList{}
+	if err := r.client.List(ctx, options, classes); err != nil {
+		return reconcile.Result{Requeue: false}, errors.Wrap(IgnoreNotFound(r.client.Status().Update(ctx, claim)), errNoDefault)
+	}
+
+	// Check to make sure there is only one default
+	// resource class defined
+	if len(classes.Items) != 1 {
+		return reconcile.Result{Requeue: false}, errors.Wrap(IgnoreNotFound(r.client.Status().Update(ctx, claim)), errMultipleDefaults)
```

Comment:
```suggestion
		// If this is the first time we encounter multiple defaults we'll be
		// requeued implicitly due to the status update. If not, we don't
		// want to requeue because...
		claim.SetConditions(corev1alpha1.ReconcileError(errors.New(errMultipleDefaults)))
		return reconcile.Result{Requeue: false}, errors.Wrap(IgnoreNotFound(r.client.Status().Update(ctx, claim)), errUpdateClaimStatus)
```

How do you feel about the suggested pattern? It's applicable to any error triggered return from this method except for the case in which we can't get the claim under reconciliation.

I suspect your current implementation intends to return an error from the reconcile method, which would result in us logging the `errMultipleDefaults` error, but not updating the claim itself to reflect the problem. In fact due to the way `errors.Wrap()` works, it's only going to return an error if the `IgnoreNotFound(r.client.Status().Update(ctx, claim))` call fails, and that error is going to indicate the failure was due to there being multiple defaults (when it was really due to there being multiple defaults _and_ the update failing).

I don't think we can assume a developer trying to write a resource claim has access to the Crossplane pod logs, so I prefer to use `ReconcileError`, which will update the claim's conditioned status to reflect that the reconcile failed and why. This way the developer writing the claim can do `kubectl describe mycoolclaim` or `kubectl get -o yaml myclaim` to get an idea of why it's not working as expected.

Lastly, I strongly encourage that we add an explanatory comment around each requeue decision, because there's often a lot of subtleties that are not immediately obvious to the reader (e.g. implicit requeues due to updating the object under reconciliation, returning an error, etc). In this particular case it's not obvious to me why we don't want to requeue.

Reply:
Very much like this proposed change and appreciate *all* off your helpful feedback! I am going to implement this change manually in order to avoid the DCO violation corrections.